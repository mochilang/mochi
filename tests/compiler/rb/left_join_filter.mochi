type Customer {
  id: int
  name: string
}

type Order {
  id: int
  customerId: int
  total: int
}

type Info {
  orderId: int
  customerName: string
  total: int
}

let customers = [
  Customer { id: 1, name: "Alice" },
  Customer { id: 2, name: "Bob" },
  Customer { id: 3, name: "Charlie" }
]

let orders = [
  Order { id: 100, customerId: 1, total: 250 },
  Order { id: 101, customerId: 2, total: 125 },
  Order { id: 102, customerId: 1, total: 300 },
  Order { id: 103, customerId: 4, total: 80 }
]

let result = from o in orders
             left join c in customers on o.customerId == c.id
             where o.total > 200
             select Info {
               orderId: o.id,
               customerName: c.name,
               total: o.total
             }

for r in result {
  print(r.orderId, r.customerName, r.total)
}
