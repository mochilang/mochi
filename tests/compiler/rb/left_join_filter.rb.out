def _query(src, joins, opts)
  where_fn = opts['where']
  items = []
  if joins.empty?
    src.each do |v|
      row = [v]
      next if where_fn && !where_fn.call(*row)
      items << row
    end
  else
    items = src.map { |v| [v] }
    joins.each_with_index do |j, idx|
      joined = []
      jitems = j['items']
      on = j['on']
      left = j['left']
      right = j['right']
      last = idx == joins.length - 1
      if right && left
        matched = Array.new(jitems.length, false)
        items.each do |l|
          m = false
          jitems.each_with_index do |r, ri|
            keep = true
            keep = on.call(*l, r) if on
            next unless keep
            m = true
            matched[ri] = true
            row = l + [r]
            if last && where_fn && !where_fn.call(*row)
              next
            end
            joined << row
          end
          row = l + [nil]
          if left && !m
            if last && where_fn && !where_fn.call(*row)
              # skip
            else
              joined << row
            end
          end
        end
        jitems.each_with_index do |r, ri|
          next if matched[ri]
          _undef = Array.new(items[0]&.length || 0, nil)
          row = _undef + [r]
          if last && where_fn && !where_fn.call(*row)
            next
          end
          joined << row
        end
      elsif right
        jitems.each do |r|
          m = false
          items.each do |l|
            keep = true
            keep = on.call(*l, r) if on
            next unless keep
            m = true
            row = l + [r]
            if last && where_fn && !where_fn.call(*row)
              next
            end
            joined << row
          end
          unless m
            _undef = Array.new(items[0]&.length || 0, nil)
            row = _undef + [r]
            if last && where_fn && !where_fn.call(*row)
              next
            end
            joined << row
          end
        end
      else
        items.each do |l|
          m = false
          jitems.each do |r|
            keep = true
            keep = on.call(*l, r) if on
            next unless keep
            m = true
            row = l + [r]
            if last && where_fn && !where_fn.call(*row)
              next
            end
            joined << row
          end
          if left && !m
            row = l + [nil]
            if last && where_fn && !where_fn.call(*row)
              next
            end
            joined << row
          end
        end
      end
      items = joined
    end
  end
  if opts['sortKey']
    items = items.map { |it| [it, opts['sortKey'].call(*it)] }
    items.sort_by! { |p| p[1] }
    items.map!(&:first)
  end
  if opts.key?('skip')
    n = opts['skip']
    items = n < items.length ? items[n..-1] : []
  end
  if opts.key?('take')
    n = opts['take']
    items = n < items.length ? items[0...n] : items
  end
  res = []
  items.each { |r| res << opts['select'].call(*r) }
  res
end

Customer = Struct.new(:id, :name, keyword_init: true)

Order = Struct.new(:id, :customerId, :total, keyword_init: true)

Info = Struct.new(:orderId, :customerName, :total, keyword_init: true)

customers = [Customer.new(id: 1, name: "Alice"), Customer.new(id: 2, name: "Bob"), Customer.new(id: 3, name: "Charlie")]
orders = [Order.new(id: 100, customerId: 1, total: 250), Order.new(id: 101, customerId: 2, total: 125), Order.new(id: 102, customerId: 1, total: 300), Order.new(id: 103, customerId: 4, total: 80)]
result = (begin
	src = orders
	_rows = _query(src, [
		{ 'items' => customers, 'on' => ->(o, c){ (o.customerId == c.id) }, 'left' => true }
	], { 'select' => ->(o, c){ Info.new(orderId: o.id, customerName: c.name, total: o.total) }, 'where' => ->(o, c){ (o.total > 200) } })
	_rows
end)
for r in result
	puts([r.orderId, r.customerName, r.total].join(" "))
end
